File: src/main/java/com/indico/storage/UploadStream.java
package com.indico.storage;

import com.indico.IndicoClient;
import com.indico.JSON;
import com.indico.RestRequest;
import okhttp3.*;
import org.json.JSONArray;

import java.io.File;
import java.io.IOException;
import java.util.Dictionary;
import java.util.Map;
import java.util.UUID;

public class UploadStream implements RestRequest<JSONArray> {

    private IndicoClient client;
    private Map<String,byte[]> bytes;
    private String fileName;

    public UploadStream(IndicoClient client) {
        this.client = client;
    }

    /**
     * A stream to upload.
     *
     * @param stream Dictionary of string identifiers (ideally file names) to byte[] streams
     * @return UploadFile
     */
    public UploadStream byteStream(Map<String,byte[]> stream) {
        if(stream != null){
            this.bytes = stream;
        }
        return this;
    }

    /**
     * Upload files and return metadata
     *
     * @return JSONArray
     * @throws IOException
     */
    @Override
    public JSONArray call() throws IOException {
        String uploadUrl = this.client.config.getAppBaseUrl() + "/storage/files/store";

        MultipartBody.Builder multipartBody = new MultipartBody.Builder().setType(MultipartBody.FORM);

        if(this.fileName == null || this.fileName.isEmpty()){
            this.fileName = UUID.randomUUID().toString();
        }
        for (String key : this.bytes.keySet()) {
            multipartBody.addFormDataPart(key, key,
                    RequestBody.create(MediaType.parse("application/octet-stream"), this.bytes.get(key)));
        }
        MultipartBody requestBody = multipartBody.build();
        Request request = new Request.Builder()
                .url(uploadUrl)
                .post(requestBody)
                .build();

        Response result = client.okHttpClient.newCall(request).execute();
        String body = result.body().string();
        JSONArray fileMeta = new JSON(body).asJSONArray();
        result.close();
        return (JSONArray) fileMeta;
    }
}

File: src/main/java/com/indico/storage/UploadStream.kt
package com.indico.storage

import com.indico.IndicoKtorClient.config
import okhttp3.MultipartBody.Builder.setType
import okhttp3.MultipartBody.Builder.addFormDataPart
import okhttp3.MultipartBody.Builder.build
import okhttp3.Request.Builder.url
import okhttp3.Request.Builder.post
import okhttp3.Request.Builder.build
import com.indico.IndicoKtorClient.httpClient
import okhttp3.OkHttpClient.newCall
import okhttp3.Call.execute
import okhttp3.Response.body
import okhttp3.ResponseBody.string
import okhttp3.Response.close
import com.indico.IndicoKtorClient
import com.indico.JSON
import com.indico.RestRequest
import com.indico.storage.UploadStream
import java.io.IOException
import okhttp3.MultipartBody
import okhttp3.Request
import java.util.UUID
import okhttp3.RequestBody
import org.json.JSONArray

class UploadStream(private val client: IndicoKtorClient) : RestRequest<JSONArray?> {
    private var bytes: Map<String, ByteArray>? = null
    private var fileName: String? = null

    /**
     * A stream to upload.
     *
     * @param stream Dictionary of string identifiers (ideally file names) to byte[] streams
     * @return UploadFile
     */
    fun byteStream(stream: Map<String, ByteArray>?): UploadStream {
        if (stream != null) {
            bytes = stream
        }
        return this
    }

    /**
     * Upload files and return metadata
     *
     * @return JSONArray
     * @throws IOException
     */
    @Throws(IOException::class)
    override fun call(): JSONArray {
        val uploadUrl = client.config.appBaseUrl + "/storage/files/store"
        val multipartBody: Builder = Builder().setType(MultipartBody.FORM)
        if (fileName == null || fileName!!.isEmpty()) {
            fileName = UUID.randomUUID().toString()
        }
        for (key in bytes!!.keys) {
            multipartBody.addFormDataPart(
                key, key,
                RequestBody.create(bytes!![key], parse.parse("application/octet-stream"))
            )
        }
        val requestBody: MultipartBody = multipartBody.build()
        val request: Request = Builder()
            .url(uploadUrl)
            .post(requestBody)
            .build()
        val result = client.httpClient.newCall(request).execute()
        val body = result.body!!.string()
        val fileMeta = JSON(body).asJSONArray()
        result.close()
        return fileMeta
    }
}